src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs

```diff
@@ -1625,97 +1649,235 @@ internal partial class OpenIdConnectServerHandler : AuthenticationHandler<OpenId
+                {
+                    Options.Logger.WriteError("invalid client authentication.");
+    
+                    await SendErrorPayloadAsync(new OpenIdConnectMessage
+                    {
+                        Error = clientNotification.Error ?? OpenIdConnectConstants.Errors.InvalidClient,
+                        ErrorDescription = clientNotification.ErrorDescription,
+                        ErrorUri = clientNotification.ErrorUri
+                    });
+    
+                    return;
+                }
+    
+                // Reject if client authentication was skipped.
+                if (clientNotification.IsSkipped && request.IsClientCredentialsGrantType())
+                {
+                    Options.Logger.WriteError("client authentication is required for client_credentials grant type.");
+    
+                    await SendErrorPayloadAsync(new OpenIdConnectMessage
+                    {
+                        Error = OpenIdConnectConstants.Errors.InvalidGrant,
+                        ErrorDescription = "client authentication is required when using client_credentials"
+                    });
+    
+                    return;
+                }
+            }
 
-                await SendErrorPayloadAsync(new OpenIdConnectMessage {
-                    Error = OpenIdConnectConstants.Errors.InvalidGrant,
-                    ErrorDescription = "Invalid access token received"
-                });
+            // Note: use the token_type_hint parameter to look up a token
+            // See https://tools.ietf.org/html/rfc7662#section-2.1
+            AuthenticationTicket ticket = null;
+            string hint;
+            if (request.Parameters.TryGetValue(OpenIdConnectConstants.Parameters.TokenTypeHint, out hint))
+            {
+                switch (hint)
+                {
+                    case OpenIdConnectConstants.TokenHintTypes.AccessToken:
+                        ticket = await ReceiveAccessTokenAsync(request.Token, request);
+                        break;
+                    case OpenIdConnectConstants.TokenHintTypes.RefreshToken:
+                        ticket = await ReceiveRefreshTokenAsync(request.Token, request);
+                        break;
+                    case OpenIdConnectConstants.TokenHintTypes.IdToken:
+                        ticket = await ReceiveIdentityTokenAsync(request.Token, request);
+                        break;
+                }
+            }
 
-                return;
+            // Note: if the coken can't be found using the hint or hint is not sent, search all supported token types.
+            // See https://tools.ietf.org/html/rfc7662#section-2.1
+            if (ticket == null)
+            {
+                ticket = await ReceiveAccessTokenAsync(request.Token, request);
+                if (ticket != null)
+                {
+                    hint = OpenIdConnectConstants.TokenHintTypes.AccessToken;
```

Comment:
You can remove that and use the (recent) `GetUsage()` extension, that will indicate you whether the received token is an authorization code, an access token, an identity token or a refresh token.